ethers-etherscan/src/contract.rs

```diff
@@ -307,13 +331,34 @@ impl Client {
     /// # }
     /// ```
     pub async fn contract_source_code(&self, address: Address) -> Result<ContractMetadata> {
+        // apply caching
+        if let Some(ref cache) = self.cache {
+            // If this is None, then we have a cache miss
+            if let Some(src) = cache.get_source(address) {
+                // If this is None, then the contract is not verified
+                return match src {
+                    Some(src) => Ok(src),
+                    None => Err(EtherscanError::ContractCodeNotVerified(address)),
+                }
+            }
+        }
```

Comment:
This is really ugly, but the alternative was to cache `Result<T, EtherscanError>`. Not sure. Any suggestions?

Reply:
you can prob do something like `if let Some(Some(src)) = cache.get_source(address)` but nbd. feel free to do in follow up